zaza/openstack/charm_tests/heat/tests.py

```diff
@@ -0,0 +1,355 @@
+
+        # Confirm stack still exists.
+        try:
+            _stack = self.heat_client.stacks.get(STACK_NAME)
+        except Exception as e:
+            # Generally, a resource availability issue if this is hit.
+            msg = 'Failed to get heat stack: {}'.format(e)
+            logging.error(msg)
+
+        # Confirm stack name.
+        logging.info('Expected, actual stack name: {}, '
+                     '{}'.format(STACK_NAME, _stack.stack_name))
+        if STACK_NAME != _stack.stack_name:
+            msg = 'Stack name mismatch, {} != {}'.format(STACK_NAME,
+                                                         _stack.stack_name)
+            logging.error(msg)
+
+    def _stack_resource_compute(self):
+        """Confirm the stack has created a nova resource and check status."""
+        logging.info('Confirming heat stack resource status...')
+
+        # Confirm existence of a heat-generated nova compute resource.
+        _resource = self.heat_client.resources.get(STACK_NAME, RESOURCE_TYPE)
+        _server_id = _resource.physical_resource_id
+        if _server_id:
+            logging.debug('Heat template spawned nova instance, '
+                          'ID: {}'.format(_server_id))
+        else:
+            msg = 'Stack failed to spawn a nova compute resource (instance).'
+            logging.error(msg)
+
+        # Confirm nova instance reaches ACTIVE status.
+        openstack_utils.resource_reaches_status(self.nova_client.servers,
+                                                _server_id,
+                                                expected_status="ACTIVE",
+                                                msg="nova instance")
+        logging.info('Nova instance reached ACTIVE status')
+
+    def _stack_delete(self):
+        """Delete a heat stack, verify."""
+        logging.info('Deleting heat stack...')
+        openstack_utils.delete_resource(self.heat_client.stacks,
+                                        STACK_NAME, msg="heat stack")
+
+    def _image_delete(self):
+        """Delete that image."""
+        logging.info('Deleting glance image...')
+        image = self.nova_client.glance.find_image(IMAGE_NAME)
+        openstack_utils.delete_resource(self.glance_client.images,
+                                        image.id, msg="glance image")
+
+    def _keypair_delete(self):
+        """Delete that keypair."""
+        logging.info('Deleting keypair...')
+        openstack_utils.delete_resource(self.nova_client.keypairs,
+                                        KEYPAIR_NAME, msg="nova keypair")
+
+    def test_100_domain_setup(self):
+        """Run required action for a working Heat unit."""
+        # Action is REQUIRED to run for a functioning heat deployment
+        logging.info('Running domain-setup action on heat unit...')
+        unit = zaza.model.get_units(self.application_name)[0]
+        assert unit.workload_status == "active"
```

Comment:
Doesn't this assume that the 100 test is the first one (after deployment).  Better to wait for active before proceeding and let the timeout kill it if it never becomes active.

Reply:
@ajkavanagh Are you suggesting a change here? I did remove the last 2 lines in this function per your next suggestion, but not sure what else needs to be done here.

Reply:
The assert assumes that the workload status is 'active'  -- this assumes that this test immediately follows deployment; another test may have, say, changed config back and the workload may not have reached active before this test runs.  If you want to ensure that it is at the active state, then use:

    zaza.model.block_until_unit_wl_status(unit.entity_id, "active")

This is the code from 2 lines below this line.

Reply:
Got it, makes sense, thank you.